src/qibolab/pulses/shape.py

```diff
@@ -79,196 +81,109 @@ def demodulate(
-        """The envelope waveform of the i component of the pulse."""
+    amplitude: float
+    tau: float
+    """The decay rate of the first exponential function."""
+    upsilon: float
+    """The decay rate of the second exponential function."""
+    g: float = 0.1
+    """Weight of the second exponential function."""
 
-        if self.pulse:
-            num_samples = int(np.rint(self.pulse.duration * sampling_rate))
-            x = np.arange(0, num_samples, 1)
-            return (
-                self.pulse.amplitude
-                * (
-                    (np.ones(num_samples) * np.exp(-x / self.upsilon))
-                    + self.g * np.exp(-x / self.tau)
-                )
-                / (1 + self.g)
-            )
-
-        raise ShapeInitError
-
-    def envelope_waveform_q(self, sampling_rate=SAMPLING_RATE) -> Waveform:
-        """The envelope waveform of the q component of the pulse."""
-
-        if self.pulse:
-            num_samples = int(np.rint(self.pulse.duration * sampling_rate))
-            return np.zeros(num_samples)
-        raise ShapeInitError
+    def i(self, times: Times) -> Waveform:
+        """Generate a combination of two exponential decays."""
+        return (
+            self.amplitude
+            * (np.exp(-times / self.upsilon) + self.g * np.exp(-times / self.tau))
+            / (1 + self.g)
+        )
 
-    def __repr__(self):
-        return f"{self.name}({format(self.tau, '.3f').rstrip('0').rstrip('.')}, {format(self.upsilon, '.3f').rstrip('0').rstrip('.')}, {format(self.g, '.3f').rstrip('0').rstrip('.')})"
+    def q(self, times: Times) -> Waveform:
+        """Generate an identically null signal."""
+        return np.zeros_like(times)
 
 
-class Gaussian(PulseShape):
+@dataclass(frozen=True)
+class Gaussian(Shape):
     r"""Gaussian pulse shape.
 
     Args:
-        rel_sigma (float): relative sigma so that the pulse standard deviation (sigma) = duration / rel_sigma
+        rel_sigma (float):
 
     .. math::
 
         A\exp^{-\frac{1}{2}\frac{(t-\mu)^2}{\sigma^2}}
     """
 
-    def __init__(self, rel_sigma: float):
-        self.name = "Gaussian"
-        self.pulse: "Pulse" = None
-        self.rel_sigma: float = float(rel_sigma)
+    amplitude: float
+    mu: float
+    sigma: float
+    """Relative standard deviation.
 
-    def __eq__(self, item) -> bool:
-        """Overloads == operator."""
-        if super().__eq__(item):
-            return self.rel_sigma == item.rel_sigma
-        return False
+    The pulse standard deviation will then be `sigma = duration /
+    rel_sigma`.
+    """
 
-    def envelope_waveform_i(self, sampling_rate=SAMPLING_RATE) -> Waveform:
-        """The envelope waveform of the i component of the pulse."""
+    def i(self, times: Times) -> Waveform:
+        """Generate a Gaussian window."""
+        return self.amplitude * np.exp(-(((times - self.mu) / self.sigma) ** 2) / 2)
```

Comment:
@stavros11 and @hay-k: in order to comply with the proposed scheme, I had to add one parameter to `Gaussian`, i.e., the average. Moreover, even the `sigma` before was `rel_sigma`, because the actual parameters depended on the duration of the pulse.

Most likely, this is not the best fit for a pulse, since we will always want to have a symmetric Gaussian.

So, maybe I should reassess some assumptions, and assume that `Times` is a 1D array corresponding to the full pulse window. Otherwise, we will have to make this assumption somewhere else.

Reply:
Taking a look at the expressions for the other pulses, I believe I already got my answer...

> So, maybe I should reassess some assumptions, and assume that `Times` is a 1D array corresponding to the full pulse window.

Some pulses have to play with the individual samples (i.e. the SNZ), so there is no way I can make a fully vectorized function: `Times` ***have to*** be 1D arrays of full windows...

Reply:
> So, maybe I should reassess some assumptions, and assume that `Times` is a 1D array corresponding to the full pulse window. Otherwise, we will have to make this assumption somewhere else.

I am not sure I understood what was the problem here. Can you elaborate?

> `Times` ***have to*** be 1D arrays of full windows...

What do you mean by full window? Do you mean the lowest and highest timestamps in Times have to correspond to the boundaries of the pulse in a sequence?

Reply:
So, in my idea there should have been a pulse reference frame, described by the pulse parameters, and the `times` are coordinate values in this reference.

E.g. a `Gaussian` pulse could have been a pulse with a certain standard deviation (and in principle `mu = 0`, so no need for a further parameter), and `times` could be shuffled and of any shape. I would have just evaluated the function for those coordinates.
However, `Gaussian` is usually understood as a centered pulse, with mean equal to half the pulse duration, and stdev expressed as fraction of the duration.
For this, I need to know the duration, so I'm already breaking the idea of a function independent on the pulse parameters.

Moreover, if you take the SNZ, it is described in terms of the final samples, not times: within your windows, you have 5 regions, and the 2nd and 4th have to consist of a single sample, that has a value different from the others.
![image](https://github.com/qiboteam/qibolab/assets/10830876/447e4958-4039-48b8-9153-b474bb5925cf)
this I can not interpret as just a function of time, because I need to know how many samples you're taking before deciding the value of the individual sample...

Reply:
I see, thanks for the details.

> a function independent on the pulse parameters

When I think of _shape_, I imagine something like this as well. I.e. you have some abstract description of shape that you can stretch and squeeze (both along time axis to accommodate various pulse durations, and along amplitude axis to achieve different scales). So, in this picture the duration is not part of the shape description. 

On the other hand, when I think of _envelope_, I imagine a more specific thing. Namely, it is tightly coupled to the time axis. I.e. an envelope necessarily has the pulse duration in it, and cannot be stretched/squeezed along time axis. It can only be offset along the time axis. In fact, the duration of envelope defines the duration of pulse, not the other way around. The duration fundamentally belongs to the envelope and not to the pulse - the pulse just "inherits" it. 

I think, the second picture is more natural and useful.

Reply:
For my personal sensitivity, `Shape` and `Envelope` are just synonyms (in this context). But it costs me nothing to change word, I'll do it.

The problem is that we'd like to have something like a duration sweeper, and this would require dilating the pulse. But while for Gaussian-like envelopes this only means changing the scale, for SNZ this is not the case: if I stretch one sample it could span two or more samples, but by definition it has to be one.

That's why I really need to change signature, and be aware of the whole window (and each individual sample).

Reply:
I completed the transition in f3aa863 f64cffc 7836f54